modules/cas_cache/layer_cache_management.c

```diff
@@ -1263,121 +1495,99 @@ static int _cache_mngt_start_queues(ocf_cache_t cache)
 	return result;
 }
 
-struct _cache_mngt_attach_context {
-	struct completion compl;
-	int *result;
-};
-
-static void _cache_mngt_attach_complete(ocf_cache_t cache, void *priv, int error)
-{
-	struct _cache_mngt_attach_context *context = priv;
-
-	*context->result = error;
-	complete(&context->compl);
-}
-
-static int _cache_mngt_cache_priv_init(ocf_cache_t cache)
+static void init_instance_complete(struct _cache_mngt_attach_context *ctx,
+		ocf_cache_t cache)
 {
-	struct cache_priv *cache_priv;
-	uint32_t cpus_no = num_online_cpus();
+	ocf_volume_t cache_obj;
+	struct bd_object *bd_cache_obj;
+	struct block_device *bdev;
+	const char *name;
 
-	cache_priv = vzalloc(sizeof(*cache_priv) +
-			cpus_no * sizeof(*cache_priv->io_queues));
-	if (!cache_priv)
-		return -OCF_ERR_NO_MEM;
+	cache_obj = ocf_cache_get_volume(cache);
+	BUG_ON(!cache_obj);
 
-	atomic_set(&cache_priv->flush_interrupt_enabled, 1);
+	bd_cache_obj = bd_object(cache_obj);
+	bdev = bd_cache_obj->btm_bd;
 
-	ocf_cache_set_priv(cache, cache_priv);
+	/* If we deal with whole device, reread partitions */
+	if (bdev->bd_contains == bdev)
+		ioctl_by_bdev(bdev, BLKRRPART, (unsigned long)NULL);
 
-	return 0;
+	/* Set other back information */
+	name = block_dev_get_elevator_name(
+			casdsk_disk_get_queue(bd_cache_obj->dsk));
+	if (name)
+		strlcpy(ctx->cmd->cache_elevator,
+				name, MAX_ELEVATOR_NAME);
 }
 
-static void _cache_mngt_cache_priv_deinit(ocf_cache_t cache)
+static void _cache_mngt_start_complete(ocf_cache_t cache, void *priv, int error)
 {
-	struct cache_priv *cache_priv = ocf_cache_get_priv(cache);
+	struct _cache_mngt_attach_context *ctx = priv;
+	bool interrupt = false;
 
-	vfree(cache_priv);
-}
+	spin_lock(&ctx->start_sync);
+	interrupt = (ctx->result == -KCAS_ERR_WAITING_INTERRUPTED);
+	if (error || interrupt) {
```

Comment:
Suggest reworking completion functions to limit amount of code executed in completion callback and under spin lock by moving most of the final function calls (ceating exported objects etc) to a delayed work.